stdlib/Artifacts/src/Artifacts.jl

```diff
@@ -0,0 +1,642 @@
+    return artifact_name, artifact_path_tail, hash
+end
+
+"""
+    macro artifact_str(name)
+
+Macro that is used to automatically ensure an artifact is installed, and return its
+location on-disk.  Automatically looks the artifact up by name in the project's
+`(Julia)Artifacts.toml` file.  Throws an error on inability to install the requested
+artifact.  If run in the REPL, searches for the toml file starting in the current
+directory, see `find_artifacts_toml()` for more.
+
+If `name` contains a forward or backward slash, all elements after the first slash will
+be taken to be path names indexing into the artifact, allowing for an easy one-liner to
+access a single file/directory within an artifact.  Example:
+
+    ffmpeg_path = @artifact"FFMPEG/bin/ffmpeg"
+
+!!! compat "Julia 1.3"
+    This macro requires at least Julia 1.3.
+
+!!! compat "Julia 1.6"
+    Slash-indexing requires at least Julia 1.6.
+"""
+macro artifact_str(name)
+    # Find Artifacts.toml file we're going to load from
+    srcfile = string(__source__.file)
+    if ((isinteractive() && startswith(srcfile, "REPL[")) || (!isinteractive() && srcfile == "none")) && !isfile(srcfile)
+        srcfile = pwd()
+    end
+    local artifacts_toml = find_artifacts_toml(srcfile)
+    if artifacts_toml === nothing
+        error(string(
+            "Cannot locate '(Julia)Artifacts.toml' file when attempting to use artifact '",
+            name,
+            "' in '",
+            __module__,
+            "'",
+        ))
+    end
+
+    # Load Artifacts.toml at compile time, so that we don't have to use `__source__.file`
+    # at runtime, which gets stale if the `.ji` file is relocated.
+    local artifact_dict = load_artifacts_toml(artifacts_toml)
+
+    # Invalidate calling .ji file if Artifacts.toml file changes
+    Base.include_dependency(artifacts_toml)
+
+    # If `name` is a constant, we can actually load and parse the `Artifacts.toml` file now,
+    # saving the work from runtime.
+    if isa(name, AbstractString)
+        # To support slash-indexing, we need to split the artifact name from the path tail:
+        local artifact_name, artifact_path_tail, hash = artifact_slash_lookup(name, artifact_dict, artifacts_toml)
+        return quote
+            Base.invokelatest(_artifact_str, $(__module__), $(artifacts_toml), $(artifact_name), $(artifact_path_tail), $(artifact_dict), $(hash))
+        end
+    else
+        return quote
+            local artifact_name, artifact_path_tail, hash = artifact_slash_lookup($(esc(name)), $(artifact_dict), $(artifacts_toml))
+            Base.invokelatest(_artifact_str, $(__module__), $(artifacts_toml), artifact_name, artifact_path_tail, $(artifact_dict), hash)
+        end
+    end
+end
+
+# Support `AbstractString`s, but avoid compilers needing to track backedges for callers
+# of these functions in case a user defines a new type that is `<: AbstractString`
+parse_toml(path::AbstractString) = parse_toml(string(path))
```

Comment:
```suggestion
parse_toml(path::AbstractString) = parse_toml(string(path)::String)
```

are needed.

But perhaps it is enough to restrict to `Union{String, SubString{String}}`?